12-offer-encoding.md

```diff
@@ -0,0 +1,840 @@
+    - otherwise:
+      - MUST fail the request if there is an `invreq_quantity` field.
+    - if `offer_amount` is present:
+      - MUST calculate the *expected amount* using the `offer_amount`:
+        - if `offer_currency` is not the `invreq_chain` currency, convert to the
+          `invreq_chain` currency.
+        - if `invreq_quantity` is present, multiply by `invreq_quantity`.`quantity`.
+      - if `invreq_amount` is present:
+        - MUST fail the request if `invreq_amount`.`msat` is less than the *expected amount*.
+        - MAY fail the request if `invreq_amount`.`msat` greatly exceeds the *expected amount*.
+    - otherwise (no `offer_amount`):
+      - MUST fail the request if it does not contain `invreq_amount`.
+    - SHOULD send an invoice in response using the `onionmsg_tlv` `reply_path`.
+  - otherwise (no `offer_node_id`, not a response to our offer):
+    - MUST fail the request if any of the following are present:
+      - `offer_chains`, `offer_features` or `offer_quantity_max`.
+    - MUST fail the request if `invreq_amount` is not present.
+    - MAY use `offer_amount` (or `offer_currency`) for informational display to user.
+    - if it sends an invoice in response:
+      - MUST use `offer_paths` if present, otherwise MUST use `invreq_payer_id` as the node id to send to.
+  - if `invreq_chain` is not present:
+    - MUST fail the request if bitcoin is not a supported chain.
+  - otherwise:
+    - MUST fail the request if `invreq_chain`.`chain` is not a supported chain.
+
+
+## Rationale
+
+`invreq_metadata` might typically contain information about the derivation of the
+`invreq_payer_id`.  This should not leak any information (such as using a simple
+BIP-32 derivation path); a valid system might be for a node to maintain a base
+payer key and encode a 128-bit tweak here.  The payer_id would be derived by
+tweaking the base key with SHA256(payer_base_pubkey || tweak).  It's also
+the first entry (if present), ensuring an unpredictable nonce for hashing.
+
+`invreq_payer_note` allows you to compliment, taunt, or otherwise engrave
+graffiti into the invoice for all to see.
+
+Users can give a tip (or obscure the amount sent) by specifying an
+`invreq_amount` in their invoice request, even though the offer specifies an
+`offer_amount`.  The recipient will only accept this if
+the invoice request amount exceeds the amount it's expecting (i.e. its
+`offer_amount` after any currency conversion, multiplied by `invreq_quantity`, if
+any).
+
+Non-offer-response `invoice_request`s are currently required to
+explicitly state the `invreq_amount` in the chain currency,
+so `offer_amount` and `offer_currency` are redundant (but may be
+informative for the payer to know how the sender claims
+`invreq_amount` was derived).
+
+# Invoices
+
+Invoices are a payment request, and when the payment is made, 
+it can be combined with the invoice to form a cryptographic receipt.
```

Comment:
This is unclear to me, what I need to combine for get the cryptographic receipt?

Reply:
It's exactly the same as a Bolt 11 invoice:
By signing the invoice, the seller commits to the description and payment hash contained in the invoice. So if the payer can produce the payment preimage it proves that the item described was paid for. In that sense, the pair (invoice, payment preimage) acts as a receipt.
Bolt 12 adds some improvements to this scheme (but the basics remain the same as Bolt 11):
- With Bolt 11 someone can prove that the invoice was paid but not that they are the one who paid it because other nodes may know the preimage too (nodes on the payment route for instance), with Bolt 12 there is a payer key that the payer can use that they are the one who paid.
- Because of the merkle tree used to sign the invoice, it's possible to reveal only parts of the invoice. For instance you can prove that you paid for "one delicious coffee" without revealing the amount that you paid for it.

Reply:
The sentence would be clearer if "it" was replaced with "payment preimage".